File: android/expoview/src/main/java/host/exp/exponent/notifications/schedulers/SchedulerImpl.java
package host.exp.exponent.notifications.schedulers;

import android.content.Context;

import java.util.HashMap;

import host.exp.exponent.kernel.ExperienceKey;
import host.exp.exponent.notifications.ExponentNotificationManager;
import host.exp.exponent.notifications.exceptions.UnableToScheduleException;

public class SchedulerImpl implements Scheduler {

  private SchedulerModel mSchedulerModel;

  private Context mApplicationContext;

  public SchedulerImpl(SchedulerModel schedulerModel) {
    this.mSchedulerModel = schedulerModel;
  }

  @Override
  public void schedule(String action) throws UnableToScheduleException {
    if (!mSchedulerModel.shouldBeTriggeredByAction(action)) {
      return;
    }
    long nextAppearanceTime = 0;

    try {
      nextAppearanceTime = mSchedulerModel.getNextAppearanceTime();
    } catch (IllegalArgumentException e) {
      throw new UnableToScheduleException();
    }

    ExperienceKey experienceKey = mSchedulerModel.getOwnerExperienceKey();
    int notificationId = mSchedulerModel.getNotificationId();
    HashMap<String, Object> details = mSchedulerModel.getDetails();

    try {
      getManager().schedule(experienceKey, notificationId, details, nextAppearanceTime, null);
    } catch (ClassNotFoundException e) {
      e.printStackTrace();
    }
  }

  @Override
  public String getIdAsString() {
    return mSchedulerModel.getIdAsString();
  }

  @Override
  public ExperienceKey getOwnerExperienceKey() {
    return mSchedulerModel.getOwnerExperienceKey();
  }

  @Override
  public void cancel() {
    ExperienceKey experienceKey = mSchedulerModel.getOwnerExperienceKey();
    int notificationId = mSchedulerModel.getNotificationId();
    getManager().cancel(experienceKey, notificationId);
  }

  @Override
  public boolean canBeRescheduled() {
    return mSchedulerModel.canBeRescheduled();
  }

  @Override
  public String saveAndGetId() {
    return mSchedulerModel.saveAndGetId();
  }

  @Override
  public void setApplicationContext(Context context) {
    mApplicationContext = context.getApplicationContext();
  }

  @Override
  public void remove() {
    cancel();
    mSchedulerModel.remove();
  }

  private ExponentNotificationManager getManager() {
    return new ExponentNotificationManager(mApplicationContext);
  }
}

File: apps/expo-go/android/expoview/src/main/java/host/exp/exponent/notifications/schedulers/SchedulerImpl.kt
package host.exp.exponent.notifications.schedulers

import android.content.Context
import host.exp.exponent.notifications.exceptions.UnableToScheduleException
import host.exp.exponent.kernel.ExperienceKey
import host.exp.exponent.notifications.ExponentNotificationManager
import java.lang.IllegalArgumentException

class SchedulerImpl(private val schedulerModel: SchedulerModel) : Scheduler {
  private var applicationContext: Context? = null

  @Throws(UnableToScheduleException::class)
  override fun schedule(action: String?) {
    if (!schedulerModel.shouldBeTriggeredByAction(action)) {
      return
    }
    val nextAppearanceTime = try {
      schedulerModel.nextAppearanceTime
    } catch (e: IllegalArgumentException) {
      throw UnableToScheduleException()
    }
    val experienceKey = schedulerModel.ownerExperienceKey
    val notificationId = schedulerModel.notificationId
    val details = schedulerModel.getDetailsMap()

    try {
      manager.schedule(experienceKey, notificationId, details, nextAppearanceTime, null)
    } catch (e: ClassNotFoundException) {
      e.printStackTrace()
    }
  }

  override val idAsString: String
    get() = schedulerModel.idAsString
  override val ownerExperienceKey: ExperienceKey
    get() = schedulerModel.ownerExperienceKey

  override fun cancel() {
    val experienceKey = schedulerModel.ownerExperienceKey
    val notificationId = schedulerModel.notificationId
    manager.cancel(experienceKey, notificationId)
  }

  override fun canBeRescheduled(): Boolean {
    return schedulerModel.canBeRescheduled()
  }

  override fun saveAndGetId(): String {
    return schedulerModel.saveAndGetId()
  }

  override fun setApplicationContext(context: Context) {
    applicationContext = context.applicationContext
  }

  override fun remove() {
    cancel()
    schedulerModel.remove()
  }

  private val manager: ExponentNotificationManager
    get() = ExponentNotificationManager(applicationContext!!)
}
